packages/astro/src/container/pipeline.ts

```diff
@@ -77,22 +77,26 @@ export class ContainerPipeline extends Pipeline {
 		let foundRoute: RouteData | undefined;
 		// options.manifest is the actual type that contains the information
 		let finalUrl: URL | undefined = undefined;
+		if (payload instanceof URL) {
+			finalUrl = payload;
+		} else if (payload instanceof Request) {
+			finalUrl = new URL(payload.url);
+		} else {
+			finalUrl = new URL(payload, new URL(request.url).origin);
+		}
 		for (const route of this.manifest.routes) {
-			if (payload instanceof URL) {
-				finalUrl = payload;
-			} else if (payload instanceof Request) {
-				finalUrl = new URL(payload.url);
-			} else {
-				finalUrl = new URL(payload, new URL(request.url).origin);
-			}
 			if (route.routeData.pattern.test(decodeURI(finalUrl.pathname))) {
 				foundRoute = route.routeData;
 				break;
-			} else if (finalUrl.pathname === '/404') {
+			}
+		}
+
+		if (!foundRoute) {
+			if (finalUrl.pathname.includes('/404')) {
```

Comment:
should this be `startsWith` instead?

Reply:
I think `/404` might be used with locale, e.g. `/de/404` -> 404 + `Astro.currentLocale == de`

Reply:
Good catch Matthew, I should update this pipeline and use the same logic of other pipelines.

Reply:
> I think `/404` might be used with locale, e.g. `/de/404` -> 404 + `Astro.currentLocale == de`

This isn't allowed, a 404 is always at root level.

What you want is to "mark" a route, and assign a locale, which is a feature we don't have yet. However you can use the middleware+locals

Reply:
That's not a special route though, is it?